src/mito2/src/read/merge.rs

```diff
@@ -0,0 +1,485 @@
+// Copyright 2023 Greptime Team
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+//! Merge reader implementation.
+
+use std::cmp::Ordering;
+use std::collections::BinaryHeap;
+use std::mem;
+
+use async_trait::async_trait;
+
+use crate::error::Result;
+use crate::read::{Batch, BatchReader, BoxedBatchReader, Source};
+
+/// Reader to merge sorted batches.
+///
+/// The merge reader merges [Batch]es from multiple sources that yields sorted batches.
+/// Batches from each source **must** obey the following rules:
+/// 1. Batch is ordered by primary key, time index, sequence desc, op type desc (we can
+/// ignore op type as sequence is already unique).
+/// 2. Batch doesn't have duplicate elements (element with same key).
+pub struct MergeReader {
+    /// Holds a min-heap for all [Node]s. Each node yields batches from a `source`.
+    ///
+    /// `Node` in this heap **must** not be EOF.
+    nodes: BinaryHeap<Node>,
+    /// Batches for the next primary key.
+    batch_merger: BatchMerger,
+}
+
+#[async_trait]
+impl BatchReader for MergeReader {
+    async fn next_batch(&mut self) -> Result<Option<Batch>> {
+        // Collect batches from sources for the same primary key and return
+        // the collected batch.
+        while !self.nodes.is_empty() {
+            // Peek current key.
+            let Some(current_key) = self.batch_merger.primary_key() else {
+                // The merger is empty, we could push it directly.
+                self.take_batch_from_heap().await?;
+                // Try next node.
+                continue;
+            };
+            // If next node has a different key, we have finish collecting current key.
+            // Safety: node is not empty.
+            if self.nodes.peek().unwrap().primary_key() != current_key {
+                break;
+            }
+            // They have the same primary key, we could take it and try next node.
+            self.take_batch_from_heap().await?;
+        }
+
+        // Merge collected batches.
+        self.batch_merger.merge_batches()
```

Comment:
Looks like `MergeReader` will also do merge/sort/dedup when TimeSeriesMemtable has done merge/sort/dedup. Maybe we can find a fast-path when batches from memtable do not overlap with each other.